tests/unit/test_haproxy_route_lib.py

```diff
@@ -0,0 +1,342 @@
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Unit tests for haproxy-route interface library."""
+
+import json
+import logging
+from typing import Any
+from unittest.mock import MagicMock
+
+import ops
+import pytest
+from charms.haproxy.v0.haproxy_route import (
+    DataValidationError,
+    HaproxyRouteProviderAppData,
+    HaproxyRouteRequirer,
+    HaproxyRouteRequirerData,
+    HaproxyRouteRequirersData,
+    LoadBalancingAlgorithm,
+    RequirerApplicationData,
+    RequirerUnitData,
+    ServerHealthCheck,
+)
+from ops.testing import Harness
+from pydantic import ValidationError
+
+logger = logging.getLogger()
+MOCK_RELATION_NAME = "haproxy-route"
+MOCK_ADDRESS = "10.0.0.1"
+
+
+@pytest.fixture(name="mock_relation_data")
+def mock_relation_data_fixture():
+    """Create mock relation data."""
+    return {
+        "service": "test-service",
+        "ports": [8080],
+        "paths": ["/api"],
+        "subdomains": ["api"],
+        "load_balancing": {"algorithm": "leastconn"},
+        "check": {"interval": 60, "rise": 2, "fall": 3, "path": "/health"},
+    }
+
+
+@pytest.fixture(name="mock_unit_data")
+def mock_unit_data_fixture():
+    """Create mock unit data."""
+    return {"address": MOCK_ADDRESS}
+
+
+@pytest.fixture(name="mock_provider_app_data")
+def mock_provider_app_data_fixture():
+    """Create mock unit data."""
+    return HaproxyRouteProviderAppData(endpoints=["https://backend.haproxy.internal/path"]).dump()
+
+
+def test_requirer_application_data_validation():
+    """
+    arrange: Create a RequirerApplicationData model with valid data.
```

Comment:
What about using test parameters instead of different tests?

Reply:
I'm closing this for now, let me know if we should move this discussion internally!